### PR TITLE
Add dynamic url system

### PR DESCRIPTION
**Pull request by 7091lapS**
Add middleware to sync url query params to the filter values in the state (thanks Shaun:))

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -38,14 +38,17 @@
...
     "jumpstate": "^2.2.2",
+    "lodash": "^4.17.4",
     "moment": "^2.18.1",
     "react": "^15.6.1",
     "react-addons-shallow-compare": "^15.6.0",
     "react-dates": "^12.2.3",
     "react-dom": "^15.6.1",
     "react-flexbox-grid": "^1.1.3",
     "react-redux": "^5.0.4",
+    "react-router-redux": "next",
```
Comment (shaundillon) [inline package.json]: I'd use a caret and the latest version here (so `^4.0.8`), then you're moving the burden of not breaking the app over to the maintainers of `react-router-redux`

File: public/js/redux/chartsRedux.js [hunk 2]
```diff
@@ -36,7 +36,7 @@ const chartsRedux = State({
 
     updateComposerVsIncopy(state, { chartData, startDate, endDate }) {
         const { composerResponse, inCopyResponse } = chartData;
-        const range = endDate.diff(startDate, 'days');
+        const range = endDate.diff(startDate, 'days') + 1;
```
Comment (akash1810) [inline public/js/redux/chartsRedux.js]: why this change?

File: public/js/helpers/routingHelpers.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
+import _set from 'lodash/set';
+import _get from 'lodash/get';
+
+function isValidValue(value) {
+    if (value === undefined || value === '') return false;
```
Comment (shaundillon) [inline public/js/helpers/routingHelpers.js]: This could be simplified to
``` js
return value !== undefined && value !== '' && Array.isArray(value) ? value.length !== 0 : true;
```

File: public/js/services/routingMiddleware.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
+import { replace } from 'react-router-redux';
+import { Actions } from 'jumpstate';
+import _isEqual from 'lodash/isEqual';
+import moment from 'moment';
```
Comment (shaundillon) [inline public/js/services/routingMiddleware.js]: is `moment`needed? [dateFns](https://date-fns.org/docs/Getting-Started) can do a lot of the same but is far smaller and supports single member imports so we can take advantage of webpack's tree-shaking.

Comment (7091lapS) [reply public/js/services/routingMiddleware.js]: So far we've been managing all dates on the app with moment objects, so this is the format they currently need to be in the state, but it seems sensible to switch to dateFns all across the app if it does everything that we use moment for! I would do that on a separate pr though.

File: public/js/redux/chartsRedux.js (hunk 2 above)
Comment (7091lapS) [reply public/js/redux/chartsRedux.js]: this is to fix the condition by which the `fillMissingDates` is called, but it's probably more correct to do change ```const composerData = composerResponse.data.length < range ?  fillMissingDates(...)``` to ```const composerData = composerResponse.data.length <= range ?  fillMissingDates(...)```
